src/lib/statistics/chunk_statistics/histograms/abstract_histogram.cpp

```diff
@@ -0,0 +1,620 @@
+   *
+   * Example:
+   *  - bin: ["intelligence", "intellij"]
+   *  - supported_characters: [a-z]
+   *  - string_prefix_length: 4
+   *  - value: intelligent
+   *
+   *  Traditionally, if we did not strip the common prefix, we would calculate the range based on the
+   *  substring of length `string_prefix_length`, which is "inte" for both lower and upper edge of the bin.
+   *  We could not make a reasonable assumption how large the share is.
+   *  Instead, we strip the common prefix ("intelli") and calculate the share based on the numerical representation
+   *  of the substring after the common prefix.
+   *  That is, what is the share of values smaller than "gent" in the range ["gence", "j"]?
+   */
+  const auto bin_min = _bin_min(bin_id);
+  const auto bin_max = _bin_max(bin_id);
+  const auto common_prefix_len = common_prefix_length(bin_min, bin_max);
+
+  DebugAssert(value.substr(0, common_prefix_len) == bin_min.substr(0, common_prefix_len),
+              "Value does not belong to bin");
+
+  const auto value_repr = _convert_string_to_number_representation(value.substr(common_prefix_len));
+  const auto min_repr = _convert_string_to_number_representation(bin_min.substr(common_prefix_len));
+  const auto max_repr = _convert_string_to_number_representation(bin_max.substr(common_prefix_len));
+  return static_cast<float>(value_repr - min_repr) / (max_repr - min_repr + 1);
+}
+
+template <typename T>
+bool AbstractHistogram<T>::_can_prune(const PredicateCondition predicate_type, const AllTypeVariant& variant_value,
+                                      const std::optional<AllTypeVariant>& variant_value2) const {
+  const auto value = type_cast<T>(variant_value);
+
+  switch (predicate_type) {
+    case PredicateCondition::Equals: {
+      const auto bin_id = _bin_for_value(value);
+      // It is possible for EqualWidthHistograms to have empty bins.
+      return bin_id == INVALID_BIN_ID || _bin_height(bin_id) == 0ul;
+    }
+    case PredicateCondition::NotEquals:
+      return min() == value && max() == value;
+    case PredicateCondition::LessThan:
+      return value <= min();
+    case PredicateCondition::LessThanEquals:
+      return value < min();
+    case PredicateCondition::GreaterThanEquals:
+      return value > max();
+    case PredicateCondition::GreaterThan:
+      return value >= max();
+    case PredicateCondition::Between: {
+      Assert(static_cast<bool>(variant_value2), "Between operator needs two values.");
+
+      if (can_prune(PredicateCondition::GreaterThanEquals, value)) {
+        return true;
+      }
+
+      const auto value2 = type_cast<T>(*variant_value2);
+      if (can_prune(PredicateCondition::LessThanEquals, value2) || value2 < value) {
+        return true;
+      }
+
+      const auto value_bin = _bin_for_value(value);
+      const auto value2_bin = _bin_for_value(value2);
+
+      // In an EqualElementCountHistogram, if both values fall into the same gap, we can prune the predicate.
```

Comment:
We are in the AbstractHistogram. Why are we talking about concrete histograms here?

Reply:
We are talking about specific implementations because estimation and pruning code is in `AbstractHistogram`, but there are some special cases that we can use for our benefit that can only occur in some histograms.
So, for example, in the line you are referring to, we are talking about gaps between bins, and only `EqualElementCountHistograms` can have gaps between bins.
These comments are supposed to help the reader understand in which situations we know for sure that no tuples satisfy the predicate, and the reader might wonder about gaps if some histograms do not have them.
On the other hand, I find that overloading these functions for two or three edge cases is not worth it.
What do you think?